File: assesment/src/scripts/pdf_parser.py
# pdf_parser.py

import os
import json
import datetime
from PyPDF2 import PdfReader

PDF_DIR = "../data/letters"
OUTPUT_FILE = "../output/parsed-pdfs.json"

def extract_year_from_filename(filename):
    try:
        year = int(filename.replace(".pdf", ""))
        if 1865 <= year <= datetime.datetime.now().year:
            return year
        else:
            raise ValueError
    except ValueError:
        raise Exception(f"Invalid filename: {filename}. Expected format: YYYY.pdf")

def clean_text(text):
    return ' '.join(text.replace('\n', ' ').split())

def parse_pdf(file_path):
    reader = PdfReader(file_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text() or ""
    
    stat = os.stat(file_path)
    return {
        "text": clean_text(text),
        "metadata": {
            "source": os.path.basename(file_path),
            "year": extract_year_from_filename(os.path.basename(file_path)),
            "pages": len(reader.pages),
            "fileSizeMB": round(stat.st_size / (1024 * 1024), 2),
            "extractedAt": datetime.datetime.now().isoformat()
        }
    }

def main():
    if not os.path.exists(PDF_DIR):
        raise Exception(f"PDF directory not found: {PDF_DIR}")

    os.makedirs(os.path.dirname(OUTPUT_FILE), exist_ok=True)

    results = []
    for filename in sorted(os.listdir(PDF_DIR)):
        if filename.endswith(".pdf"):
            file_path = os.path.join(PDF_DIR, filename)
            try:
                print(f"✅ Parsing {filename}...")
                parsed = parse_pdf(file_path)
                results.append(parsed)
            except Exception as e:
                print(f"⚠️ Failed to parse {filename}: {e}")

    with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
        json.dump(results, f, indent=2)

    print(f"\n🎉 Parsed {len(results)} files. Output saved to {OUTPUT_FILE}")

if __name__ == "__main__":
    main()

File: assesment/src/pdf-parser.ts
// src/pdf-parser.ts
import * as fs from 'fs';
import * as path from 'path';
import pdf from 'pdf-parse';

interface PDFMetadata {
  source: string;
  year: number;
  pages: number;
  fileSizeMB: number;
  extractedAt: string;
}

interface ParsedPDF {
  text: string;
  metadata: PDFMetadata;
}

const PDF_DIR = path.resolve('data/letters');
const OUTPUT_DIR = path.resolve('output');
const ALLOWED_YEAR_RANGE = { min: 1865, max: new Date().getFullYear() };

export async function parseAllPDFs(): Promise<ParsedPDF[]> {
  try {
    if (!fs.existsSync(PDF_DIR)) {
      throw new Error(`PDF directory not found: ${PDF_DIR}`);
    }
    if (!fs.existsSync(OUTPUT_DIR)) {
      fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    }

    const files = getPDFFiles(PDF_DIR);
    if (files.length === 0) {
      throw new Error(`No PDF files found in ${PDF_DIR}`);
    }

    const results: ParsedPDF[] = [];
    for (const file of files) {
      try {
        const parsed = await parsePDFFile(path.join(PDF_DIR, file));
        results.push(parsed);
        console.log(`Processed ${file} (${parsed.metadata.pages} pages)`);
      } catch (error) {
        console.error(`Failed to process ${file}:`, error instanceof Error ? error.message : error);
      }
    }

    const outputPath = path.join(OUTPUT_DIR, 'parsed-pdfs.json');
    fs.writeFileSync(outputPath, JSON.stringify(results, null, 2));
    console.log(`\n Successfully parsed ${results.length}/${files.length} PDFs. Output saved to ${outputPath}`);

    return results;
  } catch (error) {
    console.error(' Fatal error:', error instanceof Error ? error.message : error);
    process.exit(1);
  }
}

function getPDFFiles(directory: string): string[] {
  return fs.readdirSync(directory)
    .filter(file => path.extname(file).toLowerCase() === '.pdf')
    .sort();
}

async function parsePDFFile(filePath: string): Promise<ParsedPDF> {
  const buffer = fs.readFileSync(filePath);
  const { text, numpages } = await pdf(buffer);
  const filename = path.basename(filePath);

  return {
    text: cleanText(text),
    metadata: {
      source: filename,
      year: extractYearFromFilename(filename),
      pages: numpages,
      fileSizeMB: Math.round((buffer.length / (1024 * 1024)) * 100) / 100,
      extractedAt: new Date().toISOString(),
    },
  };
}

function cleanText(text: string): string {
  const preservedLineBreaks = text.replace(/([a-z])\.\n([A-Z])/g, '$1. $2');
  return preservedLineBreaks
    .replace(/\f/g, '\n')
    .replace(/\r\n/g, '\n')
    .replace(/\s+/g, ' ')
    .replace(/-\n/g, '')
    .replace(/\n{3,}/g, '\n\n')
    .trim();
}

function extractYearFromFilename(filename: string): number {
  const basename = path.basename(filename, '.pdf');
  const year = parseInt(basename);

  if (isNaN(year)) {
    throw new Error(`Invalid filename format: ${filename}. Expected YYYY.pdf`);
  }
  if (year < ALLOWED_YEAR_RANGE.min || year > ALLOWED_YEAR_RANGE.max) {
    throw new Error(`Year ${year} out of allowed range (${ALLOWED_YEAR_RANGE.min}-${ALLOWED_YEAR_RANGE.max})`);
  }

  return year;
}
